app/models/authenticator.rb

```diff
@@ -1,35 +1,39 @@
 module Authenticator
-  require_nested :Base
-  require_nested :Database
-  require_nested :Httpd
-  require_nested :Ldap
-
   def self.for(config, username = nil)
     if username == 'admin'
       Database.new(config)
     else
-      authenticator_class(config[:mode]).new(config)
+      authenticator = authenticator_class(config[:mode])
+      authenticator && authenticator.new(config)
     end
   end
 
-  def self.authenticator_class(name)
-    authenticator_classes_cache[name.to_s]
-  end
+  def self.validate_config(config)
+    return [] if config[:mode] == "none"
 
-  def self.authenticator_classes_cache
-    @authenticator_classes_cache ||= Concurrent::Hash.new do |hash, name|
-      hash[name] = authenticator_class_for(name) || force_load_authenticator_for(name)
+    authenticator = self.for(config)
+    if authenticator
+      [:mode, "authentication type, #{config[:mode].inspect}, invalid. Should be one of: #{valid_modes.join(", ")}"]
+    else
+      authenticator.validate_config
```

Comment:
Shouldn't this be the other way around?  Otherwise you'll get `undefined method 'validate_config' for 'nil'`, right?